contrib/udfs/src/main/java/org/apache/drill/exec/udfs/NearestDateUtils.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.drill.exec.udfs;
+
+import java.time.temporal.ChronoUnit;
+
+public class NearestDateUtils {
+
+  private enum timeInterval {
+    YEAR,
+    QUARTER,
+    MONTH,
+    WEEK_SUNDAY,
+    WEEK_MONDAY,
+    DAY,
+    HOUR,
+    HALF_HOUR,
+    QUARTER_HOUR,
+    MINUTE,
+    HALF_MINUTE,
+    QUARTER_MINUTE,
+    SECOND
+  }
+
+  public final static java.time.LocalDateTime getDate(java.time.LocalDateTime d, String interval) {
+    java.time.LocalDateTime newDate = d;
+    int year = d.getYear();
+    int month = d.getMonth().getValue();
+    int day = d.getDayOfMonth();
+    int hour = d.getHour();
+    int minute = d.getMinute();
+    int second = d.getSecond();
+    switch (timeInterval.valueOf(interval.toUpperCase())) {
```

Comment:
If given value is not among enum values, `valueOf` will fail. Please catch this exception and throw more user friendly `UserException.validationError()` or at least `DrillRuntimeException` where you can describe which value has been passed and which values are expected. Also please unit test that covers this functionality.